File: frontend/src/src/services/fuelLogService.js
import API from './api'; // Centralized Axios instance with JWT interceptor

export const fuelLogService = {
    // 1. Submit Driver Log
    submitDriverLog: async ({ tripId, actualDistance, refueled, litersFilled, stationName, pricePerLiter = 270.0 }) => {
        try {
            const response = await API.post('/fuel-logs/submit-driver-log', null, {
                params: {
                    tripId,
                    actualDistance,
                    refueled,
                    litersFilled: refueled ? litersFilled : 0,
                    stationName: refueled ? stationName : '',
                    pricePerLiter
                }
            });
            return response.data;
        } catch (error) {
            console.error("Error submitting driver fuel log:", error);
            throw error;
        }
    },

    // 2. Fetch All Fuel Logs
    getAllFuelLogs: async () => {
        try {
            const response = await API.get('/fuel-logs');
            return response.data;
        } catch (error) {
            console.error("Error fetching fuel logs:", error);
            throw error;
        }
    },

    // 3. Fetch Paged Fuel Logs
    getAllFuelLogsPaged: async (page = 0, size = 10) => {
        try {
            const response = await API.get('/fuel-logs/paged', {
                params: { page, size }
            });
            return response.data;
        } catch (error) {
            console.error("Error fetching paged fuel logs:", error);
            throw error;
        }
    },

    // 4. Get Fuel Log by ID
    getFuelLogById: async (id) => {
        try {
            const response = await API.get(`/fuel-logs/${id}`);
            return response.data;
        } catch (error) {
            console.error(`Error fetching fuel log #${id}:`, error);
            throw error;
        }
    },

    // 5. Get Fuel Logs by Trip ID
    getFuelLogsByTripId: async (tripId) => {
        try {
            const response = await API.get(`/fuel-logs/trip/${tripId}`);
            return response.data;
        } catch (error) {
            console.error(`Error fetching fuel logs for trip #${tripId}:`, error);
            throw error;
        }
    },

    // 6. Update Fuel Log (Admin Only)
    updateFuelLog: async (id, fuelLogDetails) => {
        try {
            const response = await API.put(`/fuel-logs/${id}`, fuelLogDetails);
            return response.data;
        } catch (error) {
            console.error(`Error updating fuel log #${id}:`, error);
            throw error;
        }
    },

    // 7. Delete Fuel Log (Admin Only)
    deleteFuelLog: async (id) => {
        try {
            const response = await API.delete(`/fuel-logs/${id}`);
            return response.data;
        } catch (error) {
            console.error(`Error deleting fuel log #${id}:`, error);
            throw error;
        }
    }
};

File: Backend/src/main/java/com/example/Logistics/service/FuelLogService.java
package com.example.Logistics.service;

import com.example.Logistics.model.FuelLog;
import com.example.Logistics.model.Trip;
import com.example.Logistics.model.TripShipment;
import com.example.Logistics.repository.FuelLogRepository;
import com.example.Logistics.repository.TripRepository;
import com.example.Logistics.repository.TripShipmentRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class FuelLogService {

    @Autowired
    private FuelLogRepository fuelLogRepository;

    @Autowired
    private TripRepository tripRepository;
    
    @Autowired
    private TripShipmentRepository tripShipmentRepository;

    // Create
    public FuelLog addFuelLog(FuelLog fuelLog) {
        if (fuelLog.getLoggedAt() == null) {
            fuelLog.setLoggedAt(LocalDateTime.now());
        }
        return fuelLogRepository.save(fuelLog);
    }

    // Read
    public List<FuelLog> getAllFuelLogs() {
        return fuelLogRepository.findAll();
    }

    public Page<FuelLog> getAllFuelLogsPaged(int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        return fuelLogRepository.findAll(pageable);
    }

    public Optional<FuelLog> getFuelLogById(Long id) {
        return fuelLogRepository.findById(id);
    }

    public List<FuelLog> getFuelLogsByTripId(Long tripId) {
        return fuelLogRepository.findByTripId(tripId);
    }

    // Update (Admin)
    public FuelLog updateFuelLog(long id, FuelLog updatedDetails) {
        FuelLog existingLog = fuelLogRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Fuel log not found with id: " + id));

        existingLog.setLitersFilled(updatedDetails.getLitersFilled());
        existingLog.setTotalAmount(updatedDetails.getTotalAmount());
        existingLog.setStationName(updatedDetails.getStationName());
        existingLog.setTrip(updatedDetails.getTrip());

        return fuelLogRepository.save(existingLog);
    }

    // Delete
    public void deleteFuelLog(long id) {
        if (!fuelLogRepository.existsById(id)) {
            throw new RuntimeException("Fuel log not found with id: " + id);
        }
        fuelLogRepository.deleteById(id);
    }

    // ⛽ Submit Driver Fuel Log (Main Method for Driver App)
    public FuelLog submitDriverFuelLog(Long tripId, double actualDistance, boolean refueled, Double litersFilled, String stationName, Double pricePerLiter) {
        
        // 🎯 Direct trip search. Agar na mile (shipment ID case), to TripShipment join table se trip nikalen.
        Trip trip = tripRepository.findById(tripId)
                .orElseGet(() -> tripShipmentRepository.findByShipmentId(tripId)
                        .map(ts -> ts.getTrip())
                        .orElseThrow(() -> new RuntimeException("Trip or Shipment relation not found with ID: " + tripId)));

        List<FuelLog> logs = fuelLogRepository.findByTripId(trip.getId());
        FuelLog fuelLog = logs.isEmpty() ? new FuelLog() : logs.get(0);

        fuelLog.setTrip(trip);
        fuelLog.setActualDistance(actualDistance);
        fuelLog.setLoggedAt(LocalDateTime.now());

        double rate = (pricePerLiter != null && pricePerLiter > 0) ? pricePerLiter : 270.0;
        double vehicleKmPerLiter = 10.0;

        double estimatedLitersConsumed = actualDistance / vehicleKmPerLiter;
        double tripBaseFuelCost = estimatedLitersConsumed * rate;

        if (refueled && litersFilled != null && litersFilled > 0) {
            fuelLog.setLitersFilled(litersFilled);
            fuelLog.setStationName(stationName != null && !stationName.trim().isEmpty() ? stationName : "Local Station");

            double midTripRefuelCost = litersFilled * rate;
            double totalTripCost = tripBaseFuelCost + midTripRefuelCost;

            fuelLog.setTotalAmount(totalTripCost);
        } else {
            fuelLog.setLitersFilled(0.0);
            fuelLog.setStationName("Existing Tank Fuel Used");
            fuelLog.setTotalAmount(tripBaseFuelCost);
        }

        return fuelLogRepository.save(fuelLog);
    }
}